File: app/src/main/java/com/kooo/evcam/v2/service/V2VhalCustomKeyObserver.java
package com.kooo.evcam.v2.service;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.kooo.evcam.VhalNative;

import io.grpc.ManagedChannel;

public final class V2VhalCustomKeyObserver {
    public interface Listener { void onCustomKeyLongPress(); }

    private static final String TAG = "V2VhalCustomKey";
    private static final int DEFAULT_SPEED_PROP_ID = 291504647;
    private static final int LONG_PRESS_VALUE = 4;
    private static final long RECONNECT_DELAY_MS = 3000;

    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final Listener listener;
    private final int buttonPropId;
    private final V2VhalStreamClient streamClient = new V2VhalStreamClient(
            TAG,
            "Channel created, session_id=",
            "Stream idle timeout (120000ms), forcing reconnect",
            "Requested all property values to stream (attempt ",
            "SendAll exhausted all retries"
    );
    private ManagedChannel channel;
    private Thread thread;
    private volatile boolean running;
    private volatile int lastButtonState = -1;
    private volatile long batchCount;
    private volatile long customKeyEventCount;

    public V2VhalCustomKeyObserver(int buttonPropId, Listener listener) {
        this.buttonPropId = buttonPropId;
        this.listener = listener;
    }

    public synchronized void start() {
        if (running) return;
        if (!VhalNative.isLibraryLoaded()) {
            Log.w(TAG, "Native library not loaded, custom key observer disabled");
            return;
        }
        running = true;
        lastButtonState = -1;
        batchCount = 0;
        customKeyEventCount = 0;
        Log.d(TAG, "configure custom key buttonPropId=" + buttonPropId);
        VhalNative.configureCustomKey(DEFAULT_SPEED_PROP_ID, buttonPropId, 0f);
        thread = new Thread(this::connectLoop, "V2VhalCustomKey");
        thread.setDaemon(true);
        thread.start();
    }

    public synchronized void stop() {
        running = false;
        disconnect();
        if (thread != null) {
            thread.interrupt();
            thread = null;
        }
    }

    private void connectLoop() {
        while (running) {
            try {
                Log.d(TAG, "Connecting to vehicle API service...");
                if (connect()) {
                    Log.d(TAG, "Connected, starting property stream");
                    streamProperties();
                }
            } catch (Throwable error) {
                Log.e(TAG, "Connection error: " + error.getMessage(), error);
            }
            disconnect();
            if (!running) break;
            try {
                Log.d(TAG, "Reconnecting in " + RECONNECT_DELAY_MS + "ms...");
                Thread.sleep(RECONNECT_DELAY_MS);
            } catch (InterruptedException ignored) {
                break;
            }
        }
    }

    private boolean connect() {
        try {
            channel = streamClient.connect().channel;
            return true;
        } catch (Throwable error) {
            Log.e(TAG, "Connect failed: " + error.getMessage(), error);
            disconnect();
            return false;
        }
    }

    private void disconnect() {
        ManagedChannel old = channel;
        channel = null;
        streamClient.disconnect(old);
    }

    private void streamProperties() throws InterruptedException {
        streamClient.streamProperties(VhalNative.getStreamMethod(), channel, new V2VhalStreamClient.Callback() {
            @Override public void onBatch(byte[] data) { processPropertyBatch(data); }
            @Override public void onStreamCompleted() { Log.d(TAG, "Property stream completed"); }
            @Override public void onStreamError(Throwable t) { Log.e(TAG, "Property stream error: " + t.getMessage(), t); }
        });
    }

    private void processPropertyBatch(byte[] data) {
        if (!VhalNative.isLibraryLoaded()) {
            Log.w(TAG, "Native library not loaded, skipping property batch processing");
            return;
        }
        int[] events;
        try { events = VhalNative.decode(data); } catch (Throwable error) { Log.e(TAG, "Failed to decode property batch: " + error.getMessage(), error); return; }
        batchCount++;
        if (events == null || events.length < 1) {
            Log.w(TAG, "Decoded empty custom key batch count=" + batchCount + " bytes=" + data.length);
            return;
        }
        int count = events[0];
        if (count <= 0 && batchCount <= 5) Log.d(TAG, "Decoded custom key batch without events bytes=" + data.length);
        for (int i = 0; i < count; i++) {
            int offset = 1 + i * 3;
            if (offset + 2 >= events.length) break;
            int type = events[offset];
            int p1 = events[offset + 1];
            int p2 = events[offset + 2];
            if (type == VhalNative.EVT_CUSTOM_KEY) {
                customKeyEventCount++;
                Log.d(TAG, "Decoded custom key event count=" + customKeyEventCount + " p1=" + p1 + " p2=" + p2 + " last=" + lastButtonState);
                handleButtonState(p1, p2);
            }
        }
    }

    private void handleButtonState(int state, int extra) {
        if (state == LONG_PRESS_VALUE && lastButtonState != LONG_PRESS_VALUE) {
            Log.d(TAG, "Custom key long press triggered, value=" + state + " extra=" + extra + " events=" + customKeyEventCount);
            mainHandler.post(listener::onCustomKeyLongPress);
        } else if (state != lastButtonState) {
            Log.d(TAG, "Custom key state changed " + lastButtonState + " -> " + state + " extra=" + extra);
        }
        lastButtonState = state;
    }

}

File: app/src/main/kotlin/com/kooo/evcam/v2/service/vhal/V2VhalCustomKeyObserver.kt
package com.kooo.evcam.v2.service.vhal

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.kooo.evcam.VhalNative

class V2VhalCustomKeyObserver(
    private val buttonPropId: Int,
    private val listener: Listener,
) {
    fun interface Listener {
        fun onCustomKeyValue4()
    }

    private val mainHandler = Handler(Looper.getMainLooper())
    @Volatile private var running = false
    private val batchListener = V2VhalSharedStream.BatchListener { data ->
        if (running) processPropertyBatch(data)
    }
    @Volatile private var lastButtonState = -1
    @Volatile private var batchCount = 0L
    @Volatile private var emptyBatchCount = 0L
    @Volatile private var customKeyEventCount = 0L

    @Synchronized
    fun start() {
        if (running) return
        if (!VhalNative.isLibraryLoaded()) {
            Log.w(TAG, "Native library not loaded, custom key observer disabled")
            return
        }
        running = true
        lastButtonState = -1
        batchCount = 0
        emptyBatchCount = 0
        customKeyEventCount = 0
        Log.d(TAG, "configure custom key buttonPropId=$buttonPropId")
        VhalNative.configureCustomKey(DEFAULT_SPEED_PROP_ID, buttonPropId, 0f)
        V2VhalSharedStream.get().register(batchListener)
        Log.d(TAG, "registered shared VHAL stream listener")
    }

    @Synchronized
    fun stop() {
        running = false
        V2VhalSharedStream.get().unregister(batchListener)
        Log.d(TAG, "unregistered shared VHAL stream listener")
    }

    private fun processPropertyBatch(data: ByteArray) {
        if (!running) return
        if (!VhalNative.isLibraryLoaded()) {
            Log.w(TAG, "Native library not loaded, skipping property batch processing")
            return
        }
        batchCount++
        val events = V2VhalEventDecoder.decode(data, TAG)
        if (events.isEmpty()) {
            emptyBatchCount++
            if (emptyBatchCount == 1L || emptyBatchCount % EMPTY_BATCH_LOG_INTERVAL == 0L) {
                Log.d(TAG, "Decoded empty custom key batch count=$emptyBatchCount total=$batchCount bytes=${data.size}")
            }
            return
        }
        for (event in events) {
            if (event.type == VhalNative.EVT_CUSTOM_KEY) {
                customKeyEventCount++
                Log.d(TAG, "Decoded custom key event count=$customKeyEventCount p1=${event.p1} p2=${event.p2} last=$lastButtonState")
                handleButtonState(event.p1, event.p2)
            }
        }
    }

    private fun handleButtonState(state: Int, extra: Int) {
        if (state == LONG_PRESS_VALUE && lastButtonState != LONG_PRESS_VALUE) {
            Log.d(TAG, "Custom key long press triggered, value=$state extra=$extra events=$customKeyEventCount")
            mainHandler.post(listener::onCustomKeyValue4)
        } else if (state != lastButtonState) {
            Log.d(TAG, "Custom key state changed $lastButtonState -> $state extra=$extra")
        }
        lastButtonState = state
    }

    companion object {
        private const val TAG = "V2VhalCustomKey"
        private const val DEFAULT_SPEED_PROP_ID = 291504647
        private const val LONG_PRESS_VALUE = 4
        private const val EMPTY_BATCH_LOG_INTERVAL = 1000L
    }
}
